google/cloud/storage/internal/async/connection_impl.cc

```diff
@@ -300,6 +263,64 @@ AsyncConnectionImpl::RewriteObject(RewriteObjectParams p) {
       cq_, stub_, std::move(current), std::move(p.request.impl_));
 }
 
+AsyncReaderConnectionFactory AsyncConnectionImpl::MakeReaderConnectionFactory(
+    google::cloud::internal::ImmutableOptions current,
+    google::cloud::storage_experimental::ReadObjectRequest request,
+    google::storage::v2::ReadObjectRequest proto_request) {
+  using StreamingRpc = google::cloud::internal::AsyncStreamingReadRpc<
+      google::storage::v2::ReadObjectResponse>;
+
+  // Capture the variables needed to create the
+  auto make_rpc = [stub = stub_, request = std::move(request)](
+                      CompletionQueue& cq,
+                      std::shared_ptr<grpc::ClientContext> context,
+                      google::cloud::internal::ImmutableOptions options,
+                      google::storage::v2::ReadObjectRequest const& proto)
+      -> future<StatusOr<std::unique_ptr<StreamingRpc>>> {
+    ApplyQueryParameters(*context, *options, request);
+    auto rpc = stub->AsyncReadObject(cq, std::move(context), std::move(options),
+                                     proto);
+    auto start = rpc->Start();
+    return start.then([rpc = std::move(rpc)](auto f) mutable {
+      if (f.get()) return make_ready_future(make_status_or(std::move(rpc)));
+      auto r = std::move(rpc);
+      return r->Finish().then([](auto f) {
+        auto status = f.get();
+        return StatusOr<std::unique_ptr<StreamingRpc>>(std::move(status));
+      });
+    });
+  };
+
+  auto const* caller = __func__;
+  return [caller, cq = cq_, current = std::move(current),
+          make_rpc = std::move(make_rpc),
+          proto_request = std::move(proto_request)](
+             storage::Generation generation,
+             std::int64_t received_bytes) mutable {
+    UpdateGeneration(proto_request, std::move(generation));
+    UpdateReadRange(proto_request, received_bytes);
+
+    auto transform = [current](auto f) mutable
+        -> StatusOr<
+            std::unique_ptr<storage_experimental::AsyncReaderConnection>> {
+      auto rpc = f.get();
+      if (!rpc) return std::move(rpc).status();
+      return std::unique_ptr<storage_experimental::AsyncReaderConnection>(
+          std::make_unique<AsyncReaderConnectionImpl>(current,
+                                                      *std::move(rpc)));
+    };
+
+    auto retry = retry_policy(*current);
+    auto backoff = backoff_policy(*current);
+    // Do not use `std::move(current)`, we need to keep such variables unchanged
+    // for future calls.
```

Comment:
Perhaps s/unchanged for future calls/valid for all factory invocations/?

Reply:
Done